python/pyarrow/tests/test_sparse_tensor.py

```diff
@@ -126,95 +138,106 @@ def ne(a, b):
+    assert sparse_tensor.type == arrow_type
+    assert np.array_equal(expected_data, result_data)
+    assert np.array_equal(expected_coords, result_coords)
 
     # Test from Tensor
     sparse_tensor = pa.SparseCOOTensor.from_tensor(tensor)
     repr(sparse_tensor)
-    assert sparse_tensor.type == arrow_type
     result_data, result_coords = sparse_tensor.to_numpy()
-    assert np.array_equal(data, result_data)
-    assert np.array_equal(coords, result_coords)
+    assert sparse_tensor.type == arrow_type
+    assert np.array_equal(expected_data, result_data)
+    assert np.array_equal(expected_coords, result_coords)
 
 
 @pytest.mark.parametrize('dtype_str,arrow_type', tensor_type_pairs)
-def test_sparse_tensor_csr_from_dense(dtype_str, arrow_type):
+def test_sparse_csr_matrix_from_dense(dtype_str, arrow_type):
     dtype = np.dtype(dtype_str)
-    dense_data = np.array([[1, 0, 2],
-                           [0, 0, 3],
-                           [4, 5, 6]]).astype(dtype)
-
-    data = np.array([[1], [2], [3], [4], [5], [6]])
-    indptr = np.array([0, 2, 3, 6])
-    indices = np.array([0, 2, 2, 0, 1, 2])
-    tensor = pa.Tensor.from_numpy(dense_data)
+    data = np.array([[8, 2, 5, 3, 4, 6]]).T.astype(dtype)
+    indptr = np.array([0, 2, 3, 4, 6])
+    indices = np.array([0, 2, 5, 0, 4, 5])
+    array = np.array([
+        [8, 0, 2, 0, 0, 0],
+        [0, 0, 0, 0, 0, 5],
+        [3, 0, 0, 0, 0, 0],
+        [0, 0, 0, 0, 4, 6],
+    ]).astype(dtype)
+    tensor = pa.Tensor.from_numpy(array)
 
     # Test from numpy array
-    sparse_tensor = pa.SparseCSRMatrix.from_dense_numpy(dense_data)
+    sparse_tensor = pa.SparseCSRMatrix.from_dense_numpy(array)
     repr(sparse_tensor)
     result_data, result_indptr, result_indices = sparse_tensor.to_numpy()
+    assert sparse_tensor.type == arrow_type
     assert np.array_equal(data, result_data)
     assert np.array_equal(indptr, result_indptr)
     assert np.array_equal(indices, result_indices)
 
     # Test from Tensor
     sparse_tensor = pa.SparseCSRMatrix.from_tensor(tensor)
     repr(sparse_tensor)
-    assert sparse_tensor.type == arrow_type
     result_data, result_indptr, result_indices = sparse_tensor.to_numpy()
+    assert sparse_tensor.type == arrow_type
     assert np.array_equal(data, result_data)
     assert np.array_equal(indptr, result_indptr)
     assert np.array_equal(indices, result_indices)
 
 
 @pytest.mark.parametrize('dtype_str,arrow_type', tensor_type_pairs)
-def test_sparse_tensor_coo_numpy_roundtrip(dtype_str, arrow_type):
+def test_sparse_coo_tensor_numpy_roundtrip(dtype_str, arrow_type):
     dtype = np.dtype(dtype_str)
-    data = np.array([[4], [9], [7], [5]]).astype(dtype)
-    coords = np.array([[0, 0], [3, 3], [1, 1], [0, 2]])
-    shape = (4, 4)
-    dim_names = ["x", "y"]
+    data = np.array([[1, 2, 3, 4, 5, 6, 7]]).T.astype(dtype)
+    coords = np.array([
+        [0, 0, 2, 3, 1, 3, 0],
+        [0, 2, 0, 4, 5, 5, 0],
+    ]).T
```

Comment:
This is actually something scipy coo format handles by summing two values at the same coordinates, see example [here](https://docs.scipy.org/doc/scipy/reference/generated/scipy.sparse.coo_matrix.html#scipy.sparse.coo_matrix).

Reply:
Removing it for now. @mrkn do you think COO tenors should have this scipy behavior?

Reply:
SciPy's COO matrix supports not only duplicated values, but also a non-lexicographical order index.  Such a non-canonical index can be canonicalized by SciPy's feature.  Furthermore, such a non-canonical index may be incompatible with other systems.  So we don't need to support these features, I think.

Reply:
FYI, we have [an issue](https://issues.apache.org/jira/browse/ARROW-4221) to support a non-canonical COO index.

In the first step, it is better to support containing non-canonical index as is.

Reply:
I think the non-canonical flag can be useful for sharing sparse matrices between Python systems.

Reply:
Got it. I agree with supporting non-canonical format.
I'm removing non-canonical test cases and this can be further discussed under [ARROW-4221](https://issues.apache.org/jira/browse/ARROW-4221).